src/test/rpc/Simulate_test.cpp

```diff
@@ -0,0 +1,819 @@
+            tx[sfDomain.jsonName] = newDomain;
+
+            // test with autofill
+            testTx(env, tx, testSimulation, 3);
+
+            tx[sfSigningPubKey.jsonName] = "";
+            tx[sfTxnSignature.jsonName] = "";
+            tx[sfSequence.jsonName] = 3;
+            tx[sfFee.jsonName] = "12";
+
+            // test without autofill
+            testTx(env, tx, testSimulation, 3);
+
+            // TODO: check that the ledger wasn't affected
+        }
+    }
+
+    void
+    testMultisignedBadPubKey()
+    {
+        testcase("Multi-signed transaction with a bad public key");
+
+        using namespace jtx;
+        Env env(*this);
+        static auto const newDomain = "123ABC";
+        Account const alice("alice");
+        Account const becky("becky");
+        Account const carol("carol");
+        Account const dylan("dylan");
+        env.fund(XRP(10000), alice);
+        env.close();
+
+        // set up valid multisign
+        env(signers(alice, 1, {{becky, 1}, {carol, 1}}));
+
+        {
+            auto validateOutput = [&](Json::Value resp, Json::Value& tx) {
+                auto result = resp[jss::result];
+                checkBasicReturnValidity(result, tx, 5, "20");
+
+                BEAST_EXPECTS(
+                    result[jss::engine_result] == "tefBAD_SIGNATURE",
+                    result[jss::engine_result].toStyledString());
+                BEAST_EXPECT(result[jss::engine_result_code] == -186);
+                BEAST_EXPECT(
+                    result[jss::engine_result_message] ==
+                    "A signature is provided for a non-signer.");
+
+                BEAST_EXPECT(
+                    !result.isMember(jss::meta) &&
+                    !result.isMember(jss::meta_blob));
+            };
+
+            Json::Value tx;
+
+            tx[jss::Account] = alice.human();
+            tx[jss::TransactionType] = jss::AccountSet;
+            tx[sfDomain.jsonName] = newDomain;
+            tx[sfSigners.jsonName] = Json::arrayValue;
+            {
+                Json::Value signer;
+                signer[jss::Account] = becky.human();
+                signer[jss::SigningPubKey] = strHex(dylan.pk().slice());
+                Json::Value signerOuter;
+                signerOuter[sfSigner.jsonName] = signer;
+                tx[sfSigners.jsonName].append(signerOuter);
+            }
+
+            // test with autofill
+            testTx(env, tx, validateOutput, false);
+
+            tx[sfSigningPubKey.jsonName] = "";
+            tx[sfTxnSignature.jsonName] = "";
+            tx[sfSequence.jsonName] = 5;
+            tx[sfFee.jsonName] = "20";  // also tests a non-base fee
+            tx[sfSigners.jsonName][0u][sfSigner.jsonName][jss::TxnSignature] =
+                "";
+
+            // test without autofill
+            testTx(env, tx, validateOutput);
+
+            // TODO: check that the ledger wasn't affected
+        }
+    }
+
+public:
+    void
+    run() override
+    {
```

Comment:
Please add tests that `simulate`  doesn't delete unfunded offers / expired NF token offers and trustlines. As they use some weird logic for cleanup  it would be good to tests them all here.

Reply:
Added in [661219c](https://github.com/XRPLF/rippled/pull/5069/commits/661219cf7272600ac5e35028d555c5a44c519066)

Reply:
Maybe you forgot to commit but I don't see expired NFT offer tests

Reply:
I did credentials instead - was easier to set up and easier to find tests I could borrow code from.